woocommerce/woocommerce
# Create additional download permissions on product save if needed

### All Submissions:


### Changes proposed in this Pull Request:

When a simple product with downloads gets converted into a variable product all the existing download permissions for past orders become invalid. This commit adds an extra verification procedure to the product save code:

1. Get all the existing download permissions for the product and all the children (variations).
2. For each download permission for the parent product, if there's a variation that offers the same file for download (same file URL) AND an equivalent download permission doesn't exist (equivalent means same file URL, same order id and same user id), then create it. This is done in a scheduled action.

Additionally, a new `WC_Customer_Download_Data_Store::create_from_data` method is added.

Closes https://github.com/woocommerce/woocommerce/issues/26475.

### How to test the changes in this Pull Request:

1. Start in `master`.
2. Create a simple product as "downloadable" with three downloadable files, let's call them `file1`, `file2` and `file3`.
3. Complete an order for the product, you should have access to all three files (in "My Account - Downloads").
4. Convert the product to variable, with two downloadable variations. Make one variation offer `file1` and the other one `file2` (use the same file URLs, display names for the files can be different).
5. Save the product, and verify that you have lost access to the files.
6. Switch to this branch, go to the product edit page and save without changing anything.
7. Go to _WooCommerce - Status - Scheduled Actions - Pending_, you should see one named `adjust_download_permissions`. Wait until it runs, or run it yourself.
8. Verify that you got access to `file1` and `file2` again, but not to `file3`.
9. Save the product again, and wait until the scheduled action runs again.
10. Check the contents of the `wp_woocommerce_downloadable_product_permissions` table in the database, there should be no duplicate entries.

### Other information:


### Changelog entry

> Enhancement - Create additional download permissions for simple downloadable products that are converted to variable products provided that there are variations offering the same files.
